File: leetcode/2-add-two-numbers/javascript/add-two-numbers-elementary-math.js

function ListNode(val, next) {
    this.val = (val === undefined) ? 0 : val;
    this.next = (next === undefined) ? null : next;
}

function addTwoNumbers(l1, l2) {
    const dummyHead = new ListNode(0);
    let curr = dummyHead;
    let carry = 0;
    while (l1 || l2 || carry) {
        let val1 = (l1) ? l1.val : 0;
        let val2 = (l2) ? l2.val : 0;
        let sum = carry + val1 + val2;
        carry = Math.floor(sum / 10);
        curr.next = new ListNode(sum % 10);
        curr = curr.next;
        if (l1) l1 = l1.next;
        if (l2) l2 = l2.next;
    }
    return dummyHead.next;
}

const l13 = new ListNode(3);
const l12 = new ListNode(4, l13);
const l11 = new ListNode(2, l12);

const l23 = new ListNode(4);
const l22 = new ListNode(6, l23);
const l21 = new ListNode(5, l22);

const result = addTwoNumbers(l11, l21);
let current = result;
while (current) {
    console.log(current.val);
    current = current.next;
}

File: leetcode/2-add-two-numbers/python/add_two_numbers_elementary_math.py
from typing import Optional

class ListNode(object):
    def __init__(self, val = 0, next = None):
        self.val = val
        self.next = next

class Solution:
    def add_two_numbers(
        self, l1: Optional[ListNode], l2: Optional[ListNode]
    ) -> Optional[ListNode]:
        dummy_head = ListNode(0)
        curr = dummy_head
        carry = 0
        while l1 or l2 or carry:
            l1_val = l1.val if l1 else 0
            l2_val = l2.val if l2 else 0
            column_sum = l1_val + l2_val + carry
            carry = column_sum // 10
            new_node = ListNode(column_sum % 10)
            curr.next = new_node
            curr = new_node
            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else None
        return dummy_head.next

sol = Solution()
l13 = ListNode(3)
l12 = ListNode(4, l13)
l11 = ListNode(2, l12)

l23 = ListNode(4)
l22 = ListNode(6, l23)
l21 = ListNode(5, l22)

result = sol.add_two_numbers(l11, l21)

while result:
    print(result.val)
    result = result.next